refactor app

illiapelypenko opened the pull request:
add tests
replace momentjs with date-fns
remove switch statements
replace bitBankService with class
fix imports
add axios mock
improve README
remove extra code

illiapelypenko committed:
refactor app
add tests
replace momentjs with date-fns
remove switch statements
replace bitBankService with class
fix imports
add axios mock
improve README
remove extra code

DmitryDadeco commented on src/models/BankUser.ts:
```diff
@@ -12,47 +20,47 @@ export abstract class BitBankUser implements BankUser {
+    const limits = {
+      [OperationType.CashIn]: this.cashInAmountByYear,
+      [OperationType.CashOut]: this.cashOutAmountByYear,
+    }[operation.type]
```
Немного запутано написано. Лучше разбить на 2 переменных. allLimits & limits = allLimits[type]

DmitryDadeco commented on src/models/BankUser.ts:
```diff
@@ -12,47 +20,47 @@ export abstract class BitBankUser implements BankUser {
+    let fee = percentageFee
+
+    if ('week_limit' in config) {
+      fee = this.getFeeWithLimits(operation, config as OperationConfigWithWeekLimit)
```
Чтобы избавится от as можно написать type guard https://www.typescriptlang.org/docs/handbook/advanced-types.html.

DmitryDadeco commented on src/models/BitBank.ts:
```diff
@@ -1,44 +1,36 @@
 
 export class BitBank implements Bank {
-  operations: Array<BankOperation> = []
+  cleanOperations: Array<BankOperation> = []
```
Почему clean? Можно лучше all?

illiapelypenko replied:
А скажи плз что ты под этим имел ввиду тогда? "Заведи для этого cleanOperations
https://github.com/illiapelypenko/fee-calculator/blob/master/src/models/BitBank.ts#L73"

DmitryDadeco commented on src/models/BitBank.ts:
```diff
@@ -53,24 +45,30 @@ export class BitBank implements Bank {
-      case OperationType.CashOut:
-        return await operation.user.cashOut(operation)
+    const operations = {
+      [OperationType.CashIn]: async () => await operation.user.cashIn(operation),
```
async/await в данной ситуации лишний.

DmitryDadeco commented on src/models/BitBank.ts:
```diff
@@ -53,24 +45,30 @@ export class BitBank implements Bank {
+    const userOperation = await operations[operation.type]
+
+    if (userOperation) {
+      return await userOperation()
```
В ретурне не обязательно писать await.

DmitryDadeco commented on src/models/BitBank.ts:
```diff
@@ -53,24 +45,30 @@ export class BitBank implements Bank {
+    }
 
-    this.operations = []
+    this.cleanOperations = []
```
Заведи функцию для чистки операций - cleanOperations. Ты походу меня не так понял, я имел в виду метод новый завести, а не переименовать переменную.

DmitryDadeco commented:
<img width="478" alt="image" src="https://user-images.githubusercontent.com/49192511/158318582-ae4c5794-d99c-445b-8e94-2f2bc2e8fe40.png">
Есть неиспользуемые экспорты.

DmitryDadeco commented:
По поводу твоего комментария "11) да min/max  одновременно не могут существовать в текущих конфигах, я так сделал на случай если будет новый конфиг в котором будет и мин и макс (расширяемость)" в таком случае у тебя одно перепишет другое, тут расширяемости нет. Так что лучше else if использовать

DmitryDadeco commented:
1) в Ридми нет команды установки пакетов
2) npm build не работает. Надо npm run build
3) Нет 100% покрытия.

DmitryDadeco commented:
Не положил тесты рядом с файлами на которые они пишутся

DmitryDadeco commented on __mocks__/axios.ts:
```diff
@@ -0,0 +1,38 @@
+
+const axios = {
+  get: jest.fn(url => {
+    const path = url.split(`${BASE_BIT_BANK_API_URL}`)[1] as keyof typeof configs
```
`const path: keyof typeof configs = ...`

DmitryDadeco commented on __mocks__/axios.ts:
```diff
@@ -0,0 +1,38 @@
+import { BASE_BIT_BANK_API_URL } from '../src/config'
```
А как этот мок работает? jest по умолчанию смотрит на папку __mocks__ и использует ее вместо нод модулей?

illiapelypenko replied:
Как я понял - да. Вот объяснение из документации https://jestjs.io/docs/manual-mocks#mocking-node-modules

DmitryDadeco replied:
А с fs так не прокатило?

illiapelypenko committed:
improve axios mock
add tests for BankUser
add tests for BitBank
add tests for BitBankService
improve BankUser
improve BitBank
add fs mock
improve README
add typeGuard util and test for it
